File: LeetCode/704-binary-search/Python/main.py
def search(nums, target) -> int:
    """
    :type nums: List[int]
    :type target: int
    :rtype: int
    """
    l: int = 0
    r: int = len(nums) - 1
    while (l < r):
        mid: int = l + ((r - l) // 2)
        if nums[mid] == target:
            return mid
        elif nums[mid] > target:
            r = mid - 1
        else:
            l = mid + 1
    return l if nums[l] == target else -1


if __name__ == "__main__":
    print("Binary Search")
    print(search([1, 2, 3, 4, 5], 2))
    print(search([1, 2, 3, 4, 5], 1))
    print(search([1, 2, 3, 4, 5], 3))
    print(search([1, 2, 3, 4, 5], 4))
    print(search([1, 2, 3, 4, 5], 5))
    print(search([1, 2, 3, 4, 5], 10))

File: LeetCode/704-binary-search/Rust/src/main.rs

fn main() {
    let v1 = vec![1, 2, 3, 4, 5, 6, 7];

    println!("search(v1) = {}", search(v1, 10));
}
pub fn search(nums: Vec<i32>, target: i32) -> i32 {
    let mut left = 0;
    let mut right = nums.len() - 1;
    let mut mid = right / 2;

    while left < right {
        if nums[mid] == target {
            return mid as i32;
        }
        if nums[mid] < target {
            left = mid + 1;
        } else {
            right = mid;
        }
        mid = (right + left) / 2;
    }

    return if nums[mid] == target { mid as i32 } else { -1 };
}
